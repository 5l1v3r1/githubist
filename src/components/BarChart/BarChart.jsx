// @flow

import React, { PureComponent, type Node } from 'react';

import styles from './BarChart.scss';

import BarChartItem from './Item';

type Item = {
  name: string,
  score: number,
  slug: string,
};

type Props = {
  items: Array<Item>,
  children: Item => Node,
};

type State = {
  maxScore: number,
};

class BarChart extends PureComponent<Props, State> {
  constructor(props: Props) {
    super(props);

    const { items } = props;

    const maxScore = items.length > 0 ? items[0].score : 1;

    this.state = {
      maxScore,
    };
  }

  calculatePercantage = (score: number): number => {
    const { maxScore } = this.state;

    return (score * 100) / maxScore;
  };

  render() {
    const { items, children } = this.props;

    return (
      <div className={styles.barChart}>
        {items.map((item, index) => (
          <BarChartItem
            key={item.slug}
            rank={index + 1}
            percentage={this.calculatePercantage(item.score)}
          >
            {children(item)}
          </BarChartItem>
        ))}
      </div>
    );
  }
}

export default BarChart;
