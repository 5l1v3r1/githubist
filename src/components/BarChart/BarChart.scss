@import '../../theme/theme';

.barChart {
  display: grid;
  padding: map-get($spacers, 2x-large) 0;
  grid-row-gap: map-get($spacers, normal);
  grid-template-rows: 1fr;
}

.item {
  display: flex;
  flex-direction: row;
}

.rank {
  padding: map-get($spacers, 2x-small) map-get($spacers, normal);
  background-color: $body-color;
  border-radius: px-to-rem(2px);
  color: $white;
  font-size: map-get($heading-sizes, h2);
}

.outer {
  position: relative;
  width: 100%;
  margin-left: map-get($spacers, normal);
  background-color: #8dbfc3;
  border-radius: px-to-rem(2px);
}

.bar {
  position: absolute;
  height: 100%;
  background-color: $secondary-color;
  border-radius: px-to-rem(2px);
}

.content {
  position: absolute;
  display: flex;
  width: 100%;
  height: 100%;
  align-items: center;
  margin-left: map-get($spacers, normal);
}
