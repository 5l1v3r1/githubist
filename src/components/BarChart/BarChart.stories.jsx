// @flow

import React from 'react';
import { storiesOf } from '@storybook/react';

import BarChart from './BarChart';

const items = [
  {
    name: 'Item 1',
    slug: 'item1',
    score: 100,
  },
  {
    name: 'Item 2',
    slug: 'item2',
    score: 80,
  },
  {
    name: 'Item 3',
    slug: 'item3',
    score: 10,
  },
];

storiesOf('BarChart', module).add('default', () => (
  <BarChart items={items}>{item => <div>{item.name}</div>}</BarChart>
));
