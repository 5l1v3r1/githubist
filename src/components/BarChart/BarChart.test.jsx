// @flow

import React from 'react';
import { shallow } from 'enzyme';

import BarChart from './BarChart';

const items = [
  {
    name: 'Item 1',
    slug: 'item1',
    score: 100,
  },
  {
    name: 'Item 2',
    slug: 'item2',
    score: 80,
  },
  {
    name: 'Item 3',
    slug: 'item3',
    score: 10,
  },
];

it('renders correctly', () => {
  const wrapper = shallow(<BarChart items={items}>{item => <div>{item.name}</div>}</BarChart>);

  expect(wrapper).toMatchSnapshot();
});
