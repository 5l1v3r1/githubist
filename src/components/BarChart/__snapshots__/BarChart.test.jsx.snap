// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <Item
    key="item1"
    percentage={100}
    rank={1}
  >
    <div>
      Item 1
    </div>
  </Item>
  <Item
    key="item2"
    percentage={80}
    rank={2}
  >
    <div>
      Item 2
    </div>
  </Item>
  <Item
    key="item3"
    percentage={10}
    rank={3}
  >
    <div>
      Item 3
    </div>
  </Item>
</div>
`;
